app/src/main/kotlin/co/netguru/baby/monitor/client/feature/voiceAnalysis/RecordingController.kt

```diff
@@ -0,0 +1,58 @@
+package co.netguru.baby.monitor.client.feature.voiceAnalysis
+
+import co.netguru.baby.monitor.client.common.RunsInBackground
+import co.netguru.baby.monitor.client.feature.machinelearning.MachineLearning
+import co.netguru.baby.monitor.client.feature.noisedetection.NoiseDetector
+import io.reactivex.Observable
+import java.nio.ByteBuffer
+import java.nio.ByteOrder
+import javax.inject.Inject
+
+class RecordingController @Inject constructor(
+    private val aacRecorder: AacRecorder
+) {
+    var voiceAnalysisOption: VoiceAnalysisOption = VoiceAnalysisOption.MACHINE_LEARNING
+    private var rawData = emptyArray<Byte>()
+    private var newData = emptyArray<Short>()
+
+    fun startRecording(): Observable<RecordingData> {
+        return aacRecorder.startRecording()
+            .doOnNext {
+                addData(it)
+            }
+            .map {
+                val data = when {
+                    voiceAnalysisOption == VoiceAnalysisOption.MACHINE_LEARNING &&
+                            newData.size >= MachineLearning.DATA_SIZE -> RecordingData.MachineLearning(
+                        rawData.toByteArray(),
+                        newData.toShortArray()
+                    )
+                    voiceAnalysisOption == VoiceAnalysisOption.NOISE_DETECTION &&
+                            newData.size >= NoiseDetector.DATA_SIZE -> RecordingData.NoiseDetection(
+                        newData.takeLast(NoiseDetector.DATA_SIZE).toShortArray()
+                    )
+                    else -> RecordingData.Raw(it)
+                }
+                if (newData.size >= MachineLearning.DATA_SIZE) clearRecordingData()
+                data
+            }
```

Comment:
you are operating on side effects here and sideffects are bad - ideally your code should be running fine no matter on what thread you would retrieve the result. You should look into reduce operator it will allow you to do all that while keeping everything pure and without side effects λ💚

Think it would look something like this(warning pseudo code)
```kotlin
aacRecorder.startRecording()
    .scan(Pair(emptyArray<Byte>(),emptyArray<Short>()){ acc, new->
    // warning i copied this and moved before not after maybe data size should be change? consider the logic i am not aware of
    val (newData,rawData) = if (previous.size >= MachineLearning.DATA_SIZE) {
        new
    }else{
    acc+new
    }    
}
// logic in when was skipped for simplicity 😉 
```